backvendor/git.go

```diff
@@ -0,0 +1,158 @@
+	buf, err := wt.run("rev-parse", tag)
+	if err != nil {
+		os.Stderr.Write(buf.Bytes())
+		return "", err
+	}
+	rev := strings.TrimSpace(buf.String())
+	return rev, nil
+}
+
+// gitRevSync updates the working tree to reflect the tag or revision
+// ref, using 'git checkout ...'. The working tree must not have been
+// locally modified.
+func (wt *WorkingTree) gitRevSync(ref string) error {
+	buf, err := wt.run("checkout", ref)
+	if err != nil {
+		os.Stderr.Write(buf.Bytes())
+	}
+	return err
+}
+
+// gitTimeFromRevision returns the commit timestamp for the revision
+// rev, using 'git show -s --pretty=format:%cI ...'.
+func (wt *WorkingTree) gitTimeFromRev(rev string) (time.Time, error) {
+	var t time.Time
+	buf, err := wt.run("show", "-s", "--pretty=format:%cI", rev)
+	if err != nil {
+		return t, err
+	}
+
+	t, err = time.Parse(time.RFC3339, strings.TrimSpace(buf.String()))
+	return t, err
+}
+
+// gitReachableTag returns the most recent reachable semver tag, using
+// 'git describe --tags --match=...', with match globs for tags that
+// are likely to be semvers. It returns ErrorVersionNotFound if no
+// suitable tag is found.
+func (wt *WorkingTree) gitReachableTag(rev string) (string, error) {
+	var tag string
+	for _, match := range []string{"v[0-9]*", "[0-9]*"} {
+		buf, err := wt.run("describe", "--tags", "--match="+match, rev)
+		output := strings.TrimSpace(buf.String())
+		if err == nil {
+			tag = output
+			break
+		}
+
+		// Catch failures due to not finding an appropriate tag
+		switch {
+		// fatal: no tag exactly matches ...
+		// fatal: no tags can describe ...
+		// fatal: no names found, cannot describe anything.
+		// fatal: no annotated tags can describe ...
+		case strings.HasPrefix(output, "fatal: no tag"),
+			strings.HasPrefix(output, "fatal: no names"),
+			strings.HasPrefix(output, "fatal: No annotated tag"):
+			err = ErrorVersionNotFound
+		default:
+			os.Stderr.Write(buf.Bytes())
+		}
+		return "", err
+	}
+
+	if tag == "" {
+		return "", ErrorVersionNotFound
+	}
+
+	log.Debugf("%s is described as %s", rev, tag)
+	fields := strings.Split(tag, "-")
+	if len(fields) < 3 {
+		// This matches a tag exactly (it must not be a semver tag)
+		return tag, nil
+	}
+	tag = strings.Join(fields[:len(fields)-2], "-")
+	return tag, nil
+}
+
+// gitFileHashesFromRef parses the output of 'git ls-tree -r' to
+// return the file hashes for the given tag or revision ref.
+func (wt *WorkingTree) gitFileHashesFromRef(ref string) (*FileHashes, error) {
+	buf, err := wt.run("ls-tree", "-r", ref)
+	if err != nil {
+		if strings.HasPrefix(buf.String(), "fatal: Not a valid object name ") {
+			// This is a branch name, not a tag name
+			return nil, ErrorInvalidRef
+		}
+
+		os.Stderr.Write(buf.Bytes())
+		return nil, err
+	}
+	fh := make(map[string]FileHash)
+	scanner := bufio.NewScanner(buf)
+	for scanner.Scan() {
+		line := scanner.Text()
+		fields := strings.Fields(line)
+		if len(fields) < 4 {
```

Comment:
Are there times where this is greater than 4? Perhaps this should just be `if len(fields) != 4`.

Reply:
Fixed.